README.md
# praye.js

_Lightweight and highly accurate low-level library for calculating (Islamic) prayer times._

## Usage

```typescript
import { PrayerManager, getCalculationMethod } from 'praye.js';

const prayerManager = new PrayerManager(getCalculationMethod('MWL'));

const aDate = new Date(2021, 4, 12);
const aHouse: Coordinates = [38.8976763, -77.036529, 18.0];

const times = prayerManager.getTimes(aDate, aHouse);
```


CHANGELOG.md
# Changelog

### [0.3.1](https://www.github.com/Mesteery/praye.js/compare/v0.3.0...v0.3.1) (2022-04-17)


### Bug Fixes

* **pkg:** remove useless node engine limitation ([ce5e8b9](https://www.github.com/Mesteery/praye.js/commit/ce5e8b996e292c090ba0c3e7b72fbe90a8683046))

## [0.3.0](https://www.github.com/Mesteery/praye.js/compare/v0.2.0...v0.3.0) (2022-04-17)


### ⚠ BREAKING CHANGES

* only return "incomplete data" from getCalculationMethod

### Features

* only return "incomplete data" from getCalculationMethod ([0613a55](https://www.github.com/Mesteery/praye.js/commit/0613a559ec6d662f643f6f6398c31633b9764c79))

## [0.2.0](https://www.github.com/Mesteery/praye.js/compare/v0.1.0...v0.2.0) (2021-05-13)


### ⚠ BREAKING CHANGES

* month bugs

### Bug Fixes

* month bugs ([fee939a](https://www.github.com/Mesteery/praye.js/commit/fee939a954ea512c03319e025a3854b2810fcd62))

## 0.1.0 (2021-05-13)


### Features

* init project ([fed8724](https://www.github.com/Mesteery/praye.js/commit/fed8724503f1a7eb5127ac962297a815a01178af))


src/__tests__/prayer.test.ts
import { Coordinates } from '../astronomy';
import { getCalculationMethod, PrayerManager } from '../prayer';

describe(PrayerManager.prototype.getTimes, () => {
	it('should compute the right times', () => {
		const prayerManager = new PrayerManager(getCalculationMethod('MWL'));

		const aDate = new Date(2021, 3, 12);
		const aHouse: Coordinates = [38.897_676_3, -77.036_529, 18];

		expect(prayerManager.getTimes(aDate, aHouse)).toStrictEqual({
			imsak: 8.860_089_038_173_626,
			fajr: 9.026_755_704_840_292,
			sunrise: 10.581_941_026_910_075,
			dhuhr: 17.147_080_969_043_08,
			asr: 20.831_494_870_257_075,
			sunset: 23.722_396_131_662_42,
			maghrib: 23.722_396_131_662_42,
			isha: 25.184_533_130_566_4,
			midnight: 29.152_168_579_286_247,
		});
	});
});


src/dmath.ts
function toRadians(degrees: number) {
	return degrees * (Math.PI / 180);
}

function toDegrees(radians: number) {
	return (radians * 180) / Math.PI;
}

export function sin(degrees: number): number {
	return Math.sin(toRadians(degrees));
}

export function cos(degrees: number): number {
	return Math.cos(toRadians(degrees));
}

export function tan(degrees: number): number {
	return Math.tan(toRadians(degrees));
}

export function arcsin(degrees: number): number {
	return toDegrees(Math.asin(degrees));
}

export function arccos(degrees: number): number {
	return toDegrees(Math.acos(degrees));
}

export function arccot(degrees: number): number {
	return toDegrees(Math.atan(1 / degrees));
}

export function arctan2(y: number, x: number): number {
	return toDegrees(Math.atan2(y, x));
}

export function fixAngle(angle: number): number {
	return fix(angle, 360);
}

export function fixHour(hour: number): number {
	return fix(hour, 24);
}

function fix(a: number, b: number): number {
	const fixed = a % b;
	return fixed < 0 ? fixed + b : fixed;
}


src/astronomy.ts
import { arccos, arcsin, arctan2, cos, fixAngle, fixHour, sin } from './dmath';

export type Coordinates = [latitude: number, longitude: number, altitude?: number];

export function getJulianDay(date: Date): number {
	let year = date.getFullYear();
	let month = date.getMonth() + 1;
	const day = date.getDate();

	if (month < 3) {
		year -= 1;
		month += 12;
	}

	return Math.floor(365.2425 * year + 30.6001 * month) + day + 1_721_027.5;
}

export function midDay(julianDate: number, time: number): number {
	const eqt = sunPosition(julianDate + time)[1];
	return fixHour(12 - eqt);
}

export function sunAngleTime(julianDate: number, latitude: number, angle: number, time: number, ccw: boolean): number {
	const [decl, eqt] = sunPosition(julianDate + time);
	const t = (1 / 15) * arccos((-sin(angle) - sin(decl) * sin(latitude)) / (cos(decl) * cos(latitude)));
	return fixHour(12 - eqt) + (ccw ? -t : t);
}

export function sunPosition(jd: number): [decl: number, eqt: number] {
	const D = jd - 2_451_545;

	const q = fixAngle(280.460_618_37 + 0.985_647_36 * D);
	const g = fixAngle(357.528 + 0.985_600_28 * D);
	const L = fixAngle(q + 1.915 * sin(g) + 0.02 * sin(2 * g));
	const e = 23.439 - 0.000_000_36 * D;

	const decl = arcsin(sin(e) * sin(L));
	const eqt = q / 15 - fixHour(arctan2(cos(e) * sin(L), cos(L)) / 15);
	return [decl, eqt];
}

const EARTH_RADIUS = 6_371_008.7714; // in meters

export function riseSetAngle(elevation: number): number {
	return 0.833 + arccos(EARTH_RADIUS / (EARTH_RADIUS + elevation));
}


src/prayer.ts
import { Coordinates, getJulianDay, midDay, riseSetAngle, sunAngleTime, sunPosition } from './astronomy';
import { arccot, fixHour, tan } from './dmath';

export interface CalculationType {
	/**
	 * Degrees (`angle`) or minutes (`minute`).
	 */
	type: 'angle' | 'minute';
	value: number;
}

/**
 * The midnight calculation method
 */
export const enum MidnightMethod {
	/**
	 * from Sunset to Sunrise
	 */
	Standard = 'standard',
	/**
	 * from Sunset to Fajr
	 */
	Jafari = 'jafari',
}

/**
 * The asr juristic methods
 */
export const enum AsrJuristic {
	/**
	 * factor: 1
	 */
	Standard = 1,
	/**
	 * factor: 2
	 */
	Hanafi,
}

/**
 * Represents a calculation method (parameters)
 */
export interface CalculationMethod {
	imsak?: CalculationType;
	/**
	 * angle in degrees
	 */
	fajr: number;
	/**
	 * in minutes
	 */
	dhuhr?: number;
	asr?: AsrJuristic;
	maghrib?: CalculationType;
	isha: CalculationType;
	midnight?: MidnightMethod;
}

/**
 * Represents prayer times
 */
export interface PrayerTimes {
	/**
	 * Imsak
	 */
	imsak: number;
	/**
	 * Fajr
	 */
	fajr: number;
	/**
	 * Sunrise
	 */
	sunrise: number;
	/**
	 * Dhur
	 */
	dhuhr: number;
	/**
	 * Asr
	 */
	asr: number;
	/**
	 * Sunset
	 */
	sunset: number;
	/**
	 * Maghrif
	 */
	maghrib: number;
	/**
	 * Isha
	 */
	isha: number;
	/**
	 * Middle of the night
	 */
	midnight: number;
}

/**
 * The method to use for higher latitudes
 *
 * http://praytimes.org/calculation#Higher_Latitudes
 */
export const enum HightLatMethods {
	/**
	 * Middle of the Night
	 *
	 * > *In this method, the period from sunset to sunrise is divided into two halves.
	 * The first half is considered to be the "night" and the other half as "day break".
	 * Fajr and Isha in this method are assumed to be at mid-night during the abnormal periods.*
	 * http://praytimes.org/calculation#Higher_Latitudes
	 */
	NightMiddle = 'NightMiddle',
	/*
	 * Angle-Based Method
	 *
	 * > *This is an intermediate solution, used by some recent prayer time calculators.
	 * Let α be the twilight angle for Isha, and let t = α/60.
	 * The period between sunset and sunrise is divided into t parts.
	 * Isha begins after the first part.
	 * For example, if the twilight angle for Isha is 15, then Isha begins at the end of the first quarter (15/60) of the night.
	 * Time for Fajr is calculated similarly.*
	 * http://praytimes.org/calculation#Higher_Latitudes
	 */
	AngleBased = 'AngleBased',
	/**
	 * One-Seventh of the Night
	 *
	 * > *In this method, the period between sunset and sunrise is divided into seven parts.
	 * Isha begins after the first one-seventh part, and Fajr is at the beginning of the seventh part.*
	 * http://praytimes.org/calculation#Higher_Latitudes
	 */
	OneSeventh = 'OneSeventh',
}

/**
 * Pre-defined calculation methods
 */
export const enum CalculationMethods {
	/**
	 * Muslim World League
	 */
	MWL = 'MWL',
	/**
	 * Islamic Society of North America
	 */
	ISNA = 'ISNA',
	/*
	 * Egyptian General Authority of Survey
	 */
	Egypt = 'Egypt',
	/**
	 * Umm Al-Qura University, Makkah
	 */
	Makkah = 'Makkah',
	/**
	 * University of Islamic Sciences, Karachi
	 */
	Karachi = 'Karachi',
	/**
	 * Institute of Geophysics, University of Tehran
	 */
	Tehran = 'Tehran',
	/**
	 * Shia Ithna-Ashari, Leva Institute, Qum
	 */
	Jafari = 'Jafari',
	/**
	 * Muslims of France
	 */
	MF = 'MF',
}

/**
 * Create complete calculation method
 *
 * @param method The calculation method
 * @returns The provided method with defaults
 */
export function createCalculationMethod(method: CalculationMethod): Required<CalculationMethod> {
	return {
		imsak: method.imsak ?? { type: 'minute', value: 10 },
		fajr: method.fajr,
		dhuhr: method.dhuhr ?? 0,
		asr: method.asr ?? AsrJuristic.Standard,
		maghrib: method.maghrib ?? { type: 'minute', value: 0 },
		isha: method.isha,
		midnight: method.midnight ?? MidnightMethod.Standard,
	};
}

/**
 * Get calculation method
 *
 *
 * # Example
 * ~~~~
 * import { getCalculationMethod } from 'praye.js';
 *
 * getCalculationMethod('MWL');
 * getCalculationMethod('Makkah', true); // in ramadan period
 * getCalculationMethod('ISNA');
 * ~~~~
 *
 * @param method The calculation method
 * @param isRamadan if it's in the ramadan period (only for Makkah's method), default `false`
 * @returns calculation method
 */
export function getCalculationMethod(
	method: keyof typeof CalculationMethods,
	isRamadan = false,
): CalculationMethod | undefined {
	return {
		[CalculationMethods.MWL]: { fajr: 18, isha: { type: 'angle', value: 17 } },
		[CalculationMethods.ISNA]: { fajr: 15, isha: { type: 'angle', value: 15 } },
		[CalculationMethods.Egypt]: { fajr: 19.5, isha: { type: 'angle', value: 17.5 } },
		[CalculationMethods.Makkah]: { fajr: 19.5, isha: { type: 'minute', value: isRamadan ? 120 : 90 } },
		[CalculationMethods.Karachi]: { fajr: 18, isha: { type: 'angle', value: 18 } },
		[CalculationMethods.Tehran]: {
			fajr: 17.7,
			maghrib: { type: 'angle', value: 4.5 },
			isha: { type: 'angle', value: 14 },
			midnight: MidnightMethod.Jafari,
		},
		[CalculationMethods.Jafari]: {
			fajr: 16,
			maghrib: { type: 'angle', value: 4 },
			isha: { type: 'angle', value: 14 },
			midnight: MidnightMethod.Jafari,
		},
		[CalculationMethods.MF]: { fajr: 12, isha: { type: 'angle', value: 12 } },
	}[method] as CalculationMethod | undefined;
}

function timeDiff(time1: number, time2: number): number {
	return fixHour(time2 - time1);
}

function asrTime(julianDay: number, latitude: number, factor: AsrJuristic, time: number): number {
	const decl = sunPosition(julianDay + time)[0];
	const angle = -arccot(factor + tan(Math.abs(latitude - decl)));
	return sunAngleTime(julianDay, latitude, angle, time, false);
}

/**
 * The prayer manager
 *
 * # Example
 * ~~~~
 * import { PrayerManager, getCalculationMethod } from 'praye.js';
 *
 * const prayerManager = new PrayerManager(getCalculationMethod('MWL'));
 * // With Makkah
 * const prayerManager = new PrayerManager(getCalculationMethod('Makkah', true)); // we are in Ramadan!
 * // Custom
 * const prayerManager = new PrayerManager({ fajr: Math.random() * 10 }); // don't do this
 *
 * const aDate = new Date(2021, 4, 12);
 * const aHouse = [38.8976763, -77.036529, 18];
 *
 * const prayers = prayerManager.getTimes(aDate, aHouse);
 * ~~~~
 */
export class PrayerManager {
	// eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
	#method!: Required<CalculationMethod>;

	public getCalculationMethod(): Required<CalculationMethod> {
		return this.#method;
	}

	public setCalculationMethod(method: CalculationMethod): void {
		this.#method = createCalculationMethod(method);
	}

	/**
	 * Initialize a PrayerManager
	 *
	 * @param method The calculation method to use
	 * @param hightLatMethod The high latitudes method to use
	 */
	public constructor(method: CalculationMethod, public hightLatMethod?: keyof typeof HightLatMethods) {
		this.setCalculationMethod(method);
	}

	/**
	 * The prayer manager
	 *
	 * # Example
	 * ~~~~
	 * import { PrayerManager, getPrayerMethod } from 'praye.js';
	 *
	 * const prayerManager = new PrayerManager(getPrayerMethod('MWL'));
	 *
	 * const aDate = new Date(2021, 4, 12);
	 * const aHouse = [38.8976763, -77.036529, 18];
	 * const prayers = prayerManager.getTimes(aDate, aHouse);
	 * ~~~~
	 */
	public getTimes(date: Date, [latitude, longitude, altitude = 0]: Coordinates): PrayerTimes {
		const julianDay = getJulianDay(date) - longitude / (15 * 24);
		const method = this.#method;
		const adjust = longitude / 15;

		let imsak = sunAngleTime(julianDay, latitude, method.imsak.value, 5 / 24, true) - adjust;
		let fajr = sunAngleTime(julianDay, latitude, method.fajr, 5 / 24, true) - adjust;
		const sunrise = sunAngleTime(julianDay, latitude, riseSetAngle(altitude), 6 / 24, true) - adjust;
		const dhuhr = midDay(julianDay, 12 / 24) - adjust + method.dhuhr / 60;
		const asr = asrTime(julianDay, latitude, method.asr, 13 / 24) - adjust;
		const sunset = sunAngleTime(julianDay, latitude, riseSetAngle(altitude), 18 / 24, false) - adjust;
		let maghrib = sunAngleTime(julianDay, latitude, method.maghrib.value, 18 / 24, false) - adjust;
		let isha = sunAngleTime(julianDay, latitude, method.isha.value, 18 / 24, false) - adjust;

		if (this.hightLatMethod) {
			const nightTime = timeDiff(sunset, sunrise);

			imsak = this.adjustHighlatTime(imsak, sunrise, method.imsak.value, nightTime, true);
			fajr = this.adjustHighlatTime(fajr, sunrise, method.fajr, nightTime, true);
			maghrib = this.adjustHighlatTime(maghrib, sunset, method.maghrib.value, nightTime, false);
			isha = this.adjustHighlatTime(isha, sunset, method.isha.value, nightTime, false);
		}

		if (method.imsak.type === 'minute') {
			imsak = fajr - method.imsak.value / 60;
		}
		if (method.maghrib.type === 'minute') {
			maghrib = sunset - method.maghrib.value / 60;
		}
		if (method.isha.type === 'minute') {
			isha = maghrib - method.isha.value / 60;
		}

		return {
			imsak,
			fajr,
			sunrise,
			dhuhr,
			asr,
			sunset,
			maghrib,
			isha,
			midnight:
				sunset + (method.midnight === MidnightMethod.Standard ? timeDiff(sunset, sunrise) : timeDiff(sunset, fajr)) / 2,
		};
	}

	private adjustHighlatTime(time: number, base: number, angle: number, night: number, ccw: boolean): number {
		const portion = this.nightPortion(angle, night);
		const diff = ccw ? timeDiff(time, base) : timeDiff(base, time);

		if (portion > diff) {
			return time;
		}

		return base + (ccw ? -portion : portion);
	}

	private nightPortion(angle: number, night: number): number {
		return (
			{
				[HightLatMethods.NightMiddle]: 1 / 2,
				[HightLatMethods.AngleBased]: (1 / 60) * angle,
				[HightLatMethods.OneSeventh]: 1 / 7,
			}[this.hightLatMethod!] * night
		);
	}
}
